### ISSUE TITLE
make_fx: torch.where scalar promotion burns in device

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#104146]
```
import torch
from torch.fx.experimental.proxy_tensor import make_fx

def f(x):
    return torch.where(x > 0.5, x, 1.0)
g = make_fx(f)(torch.rand(2, 3))
print(g)
```

yields
```
def forward(self, x_1):
    gt = torch.ops.aten.gt.Scalar(x_1, 0.5)
    scalar_tensor = torch.ops.aten.scalar_tensor.default(1.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'))
    where = torch.ops.aten.where.self(gt, x_1, scalar_tensor);  gt = x_1 = scalar_tensor = None
    return where
```

Note the hard-coded cpu device, which we are getting as we trace the promotion of `1.0` from float -> tensor.

We just spend almost 2 weeks debugging IMAs in distributed training that were root caused from this behavior during export :( . Now that we have root caused it, it's trivial to detect device burn-in and ask people to change the model. But in the long run we'd like to solve this entirely. Any thoughts on how to do so?

**Comment by ezyang** [comment]
The unhelpful answer is, we never gave any promise that make_fx produces device polymorphic code, in fact, explicitly everything is specialized, except for sizes when you do dynamic shapes.  Sooooo "don't assume export is device polymorphic." You can use fake tensor to trace in the "true" device even if you don't have the hardware available (well, you actually kinda need a CUDA build of PT to do fake cuda tracing but this is a bug).

The more satisfying answer is maybe we should be device polymorphic, since we are already pretty close?  To do it properly though, we have to introduce a symbolic device, and you would not be allowed to conditional on a device being cpu/cuda... but we do this a lot. Perhaps another way is to treat 'meta' as a polymorphic device: you know to replace meta with the true device as necessary.

**Comment by suo** [comment]
Yeah, I acknowledge this is an expansion of the guarantee that make_fx provides. But it is definitely a useful one—any form of single-controller distributed setup will likely want this, since we can't be burning in a specific CUDA device or whatever. 

And export will definitely want this; device burn-in is a pretty common complaint with torch.jit.trace as is. torch.export improves on `torch.jit.trace` somewhat because at least make_fx won't burn in device for `*_like` creation ops, but it still happens (like in the example here).

It's not super urgent, since we can just error out and there is a workaround where you register the scalar as a single-element tensor buffer. But it's obviously a worse experience than "it just works".

How much work would this be? Naively it seems like it wouldn't be TOO bad; the only place where practically we have specialized to device is in this scalar-promotion case.

**Comment by ezyang** [comment]
> How much work would this be? Naively it seems like it wouldn't be TOO bad; the only place where practically we have specialized to device is in this scalar-promotion case.

This depends a lot on what exactly you want to do. For example, I could imagine several distinct ways of doing this, which have varying levels of work involved:

1. Carefully rearrange our decomps so we don't ever generate graphs that have explicit device in them, if the user didn't specify so. I assume this is what you mean by "it shouldn't be too bad". If we don't actually add a symbolic device concept, we may have to add some new, unnatural operators for simultaneously doing an operation (like allocating a constant) while getting the device from another tensor. It is kind of bad to do it this way, because these operators can keep tensors live longer than they should
2. Add a concept of symbolic device and allow it to propagate through graphs. This is a pretty big lift because you have to add an entirely new SymDevice concept AND that's a new type that can flow through graphs that you have to deal with (we're still trying to handle on the places where sizes were done.)
3. Don't do anything, and instead write a graph transformation that can convert device arguments as necessary later.

> But it is definitely a useful one—any form of single-controller distributed setup will likely want this, since we can't be burning in a specific CUDA device or whatever.

It should be relatively simple to support polymorphic device index; literally just do your export on cuda device 0, and then post facto rewrite all occurrences of cuda:0 to the appropriate device type. (Things get more complicated if you end up with a multi-device trace, but you were asking for it.) I think it would be relatively simple to support this as a service on FX graphs containing only ATen ops.
